smack-core/src/main/java/org/jivesoftware/smack/packet/Stanza.java

```diff
@@ -458,6 +468,144 @@ public ExtensionElement removeExtension(ExtensionElement extension)  {
         return removeExtension(extension.getElementName(), extension.getNamespace());
     }
 
+    /**
+     * Returns <tt>true</tt> if custom attributes management is enabled.
+     * @return <tt>true</tt> if custom attributes management is enabled.
+     */
+    public static boolean isCustomAttributesEnabled() {
+        return customAttributesEnabled;
+    }
+
+    /**
+     * Enables the possibility to add custom attributes to the stanza.
+     * <br/><b>Enabling this feature breaks the XMPP standard and may induce unexpected behaviors!
+     * <br/>If your use-case allow it, prefer using ExtensionElement instead, and if not, be sure to triple-check the result.</b>
+     */
+    public static void enableCustomAttributes() {
+        customAttributesEnabled = true;
+    }
+
+    private static void requireCustomAttributesEnabled(String methodName) {
+        if (!isCustomAttributesEnabled()) {
+            throw new IllegalStateException(
+                    "You cannot use " + methodName + " without enabling it first with enableCustomAttributes. " +
+                    "Ensure you are aware of the consequences of doing so.");
+        }
+    }
+
+    /**
+     * Replaces all custom attributes of the stanza.
+     * @param attributes a map (name/value) of all the custom attributes of this stanza. SHould not be {@null}
+     * @throws IllegalStateException if custom attributes management is not enabled.
+     */
+    public void setCustomAttributes(Map<String, String> attributes) {
+        requireCustomAttributesEnabled("setCustomAttributes");
+        assert attributes != null;
+
+        synchronized (customAttributes) {
```

Comment:
I really don't think we need to design this API thread safe. I know legacy Smack code does it here and there, but we shouldn't do it in this case.